src/Compilers/CSharp/Portable/Binder/Binder.cs

```diff
@@ -773,6 +773,19 @@ internal Binder WithPatternVariablesIfAny(ArgumentListSyntax initializerArgument
             return new PatternVariableBinder(initializerArgumentListOpt, initializerArgumentListOpt.Arguments, this);
         }
 
+        internal void BuildAndAddPatternVariables(
+            ArrayBuilder<LocalSymbol> builder,
+            CSharpSyntaxNode node = null,
+            ImmutableArray<CSharpSyntaxNode> nodes = default(ImmutableArray<CSharpSyntaxNode>))
+        {
+            var patterns = PatternVariableFinder.FindPatternVariables(node, nodes);
+            foreach (var pattern in patterns)
+            {
+                builder.Add(SourceLocalSymbol.MakeLocal(ContainingMemberOrLambda, this, RefKind.None, pattern.Type, pattern.Identifier, LocalDeclarationKind.PatternVariable));
+            }
+            patterns.Free();
```

Comment:
Consider allocating `patterns` here and passing to `FindPatternVariables` as an argument, so the instance is allocated and freed in the same method.